### PR TITLE
llama : fix embeddings

### PR DESCRIPTION
**Pull request by ggerganov**
ref #5655, #5783

- `llama_batch.logits` now also indicates if embeddings are output for that token
- fix `llama_get_embeddings_ith()` to return token embeddings
- add `llama_get_embeddings_seq()` to return sequence embeddings
- fix `embedding` example to work both with BERT and non-BERT models
- fix `server` to get the resulting embeddings correctly
- add `examples/server-embd.py` helper script
- `server` supports `--pooling`

TODO:

- rename `llama_batch.logits` to `llama_batch.output` (future PR)
- [x] BERT graph does not need KV cache - remove it
- `server` should not queue partial prompts when computing embeddings (future PR)

### CODE REVIEW

**Comment by cebtenzzre** [comment]
While we're modifying embeddings.cpp - shouldn't this:
```cpp
struct llama_batch batch = llama_batch_init(n_batch, 0, n_prompts);
```
be this instead:
```cpp
struct llama_batch batch = llama_batch_init(n_batch, 0, 1);
```
Because only one sequence ID is assigned per token?

File: examples/embedding/embedding.cpp
Comment (cebtenzzre) [inline examples/embedding/embedding.cpp:1]: I am now getting only repeating `-nan` out of embedding.cpp - have you tried this?

Comment (ggerganov) [reply examples/embedding/embedding.cpp:1]: Yup, there are still some issues with how the embeddings are extracted, that I realized after opening the PR. Will try to fix tomorrow.

The `server` functionality is also broken since we currently will split the input into `n_batch`, but this does not work for embeddings because of the non-causal mask (i.e. each token has to attend to all other tokens of the input). So there should be some restriction added in the batching logic to no split individual inputs

File: llama.cpp [hunk 1]
```diff
@@ -8051,46 +8053,53 @@ static int llama_decode_internal(
...
-        ggml_backend_t embeddings_backend = ggml_backend_sched_get_node_backend(lctx.sched, embeddings);
-        ggml_backend_tensor_get_async(embeddings_backend, embeddings, embedding_out.data(), embd_pos*sizeof(float), embd_size*sizeof(float));
-        ggml_backend_synchronize(embeddings_backend);
+        if (batch.logits) {
+            embeddings_out.resize(n_embd * n_tokens);
+            for (uint32_t i = 0; i < n_tokens; i++) {
+                if (batch.logits[i] == 0) {
+                    continue;
+                }
+                ggml_backend_tensor_get_async(backend_embd, embd, embeddings_out.data() + (n_embd*i), (n_embd*i)*sizeof(float), n_embd*sizeof(float));
```
Comment (iamlemec) [inline llama.cpp:8099]: So the idea is the embeddings will end up in the position of the last token regardless of whether we do last token vs mean pooling? And logits all on lets you get every token embedding. Seems nice.

I think in the case of mean pooling, we would want to make the offset above `n_embd*batch.seq_id[i][0]`. I wonder if it would be better to just use info on pooling type to decide what to do here.

**Comment by tybalex** [comment]
should the embeddings from `server.cpp` get normalized too? [embedding.cpp](https://github.com/ggerganov/llama.cpp/blob/008f3fc7912bd8ecd01603fcbc580096123186cb/examples/embedding/embedding.cpp#L55) does that.

**Comment by iamlemec** [comment]
Regarding removing the KV cache, I think this will give a big speedup. I did a flamegraph on CUDA, and it was spending fully 50% of the time in calls to `llama_kv_cell.has_seq_id` for constructing the attention mask in `llama_set_inputs`. If you know everything is within batch though, you can construct the attention matrix from just the `batch` object.

File: llama.cpp (hunk 1 above)
Comment (ggerganov) [reply llama.cpp]: Yes, I wanted to have a more uniform API for logits and embeddings. Updated the code to take into account the pooling type

File: llama.cpp [hunk 2]
```diff
@@ -8275,46 +8280,57 @@ static int llama_decode_internal(
...
-        ggml_backend_t embeddings_backend = ggml_backend_sched_get_node_backend(lctx.sched, embeddings);
-        ggml_backend_tensor_get_async(embeddings_backend, embeddings, embedding_out.data(), embd_pos*sizeof(float), embd_size*sizeof(float));
-        ggml_backend_synchronize(embeddings_backend);
+        if (batch.logits) {
+            embeddings_out.resize(n_embd * n_tokens);
+            for (uint32_t i = 0; i < n_tokens; i++) {
+                if (batch.logits[i] == 0) {
+                    continue;
+                }
+                if (hparams.pooling_type == LLAMA_POOLING_TYPE_CLS) {
```
Comment (ggerganov) [inline llama.cpp:8326]: @iamlemec What is the meaning of `CLS` in this context? I don't associate this abbreviation with anything

Comment (cebtenzzre) [reply llama.cpp]: I believe this mode uses the embedding of the CLS token (which is the Bert equivalent of BOS) instead of averaging the embedding of all tokens.

Comment (iamlemec) [reply llama.cpp]: Yeah, always confused me since I think CLoSe, but I guess it stands for (sentence) classification. There's something to be said for replacing it (and the command line flag) with something more expressive like START.

File: examples/server-embd.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+result = []
+
+async def requests_post_async(*args, **kwargs):
+    return await asyncio.to_thread(requests.post, *args, **kwargs)
+
+async def main():
+    model_url = "http://127.0.0.1:6900"
+    responses: list[requests.Response] = await asyncio.gather(*[requests_post_async(
+        url= f"{model_url}/embedding",
+        json= {"content": str(0)*32}
```
Comment (cebtenzzre) [inline examples/server-embd.py:16]:
```suggestion
        json= {"content": "0"*32}
```

File: llama.cpp [hunk 4]
```diff
@@ -7999,12 +8031,37 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
                         (hparams.causal_attn && lctx.kv_self.cells[i].pos > pos)) {
```
Comment (iamlemec) [inline llama.cpp:8031]: First clause is redundant here.

**Comment by cebtenzzre** [comment]
With these changes I'm getting an MSE of 1.3e-3 relative to Sentence Transformers on WikiText with nomic-embed-text-v1.f16.gguf instead of the previous 5.62e-10. Not sure what the problem is.

**Comment by ggerganov** [comment]
> With these changes I'm getting an MSE of 1.3e-3 relative to Sentence Transformers on WikiText with nomic-embed-text-v1.f16.gguf instead of the previous 5.62e-10. Not sure what the problem is.

Does it use mean pooling? I think I got it wrong again - checking

File: llama.cpp [hunk 5]
```diff
@@ -8275,46 +8336,64 @@ static int llama_decode_internal(
...
+            embeddings_out.resize(n_embd * n_tokens);
+            for (uint32_t i = 0; i < n_tokens; i++) {
+                if (batch.logits[i] == 0) {
+                    continue;
+                }
+                switch (hparams.pooling_type) {
+                    case LLAMA_POOLING_TYPE_CLS:
+                        ggml_backend_tensor_get_async(backend_embd, embd, embeddings_out.data() + (n_embd*i), (n_embd*batch.seq_id[i][0])*sizeof(float), n_embd*sizeof(float));
+                        break;
+                    case LLAMA_POOLING_TYPE_MEAN:
```
Comment (iamlemec) [inline llama.cpp:8386]: We should have the `LLAMA_POOLING_TYPE_MEAN` case join the `LLAMA_POOLING_TYPE_CLS` case due to the output order of the averaging matrix.

Comment (ggerganov) [reply llama.cpp:8386]: Yes. I keep getting confused with the sequence-based instead of token-based embedding extraction.

Will try to modify the API to make things more clear

Comment (cebtenzzre) [reply llama.cpp:8386]: Cool, this change fixes the NaNs from embedding.cpp, and results in the MSE of nomic-embed-text-v1.f16.gguf actually being lower than before (4.71e-10 vs 5.62e-10).

**Comment by ggerganov** [comment]
So I've updated the API to support both sequence and token embeddings:

https://github.com/ggerganov/llama.cpp/blob/79e4eede2374870e915af4fe03a291a702879e52/llama.h#L657-L671

The sequence embeddings (i.e. `llama_embeddings_seq`) work only when `pooling_type != LLAMA_POOLING_TYPE_NONE`, while token embeddings are populated when `pooling_type == LLAMA_POOLING_TYPE_NONE`. In the future, we can think about adding pooling support for all models (not just BERT) - not sure if it would be useful, but it fits into the implementation logic

With this change, to make the `embedding` example work with both token and sequence emebeddings (i.e. with and without pooling), we do the following:

https://github.com/ggerganov/llama.cpp/blob/fc9af156ffa5b795bafc871554fee3ccfb800c34/examples/embedding/embedding.cpp#L52-L65

First try to get sequence embeddings. This would return NULL if pooling is disabled and we fallback to token embeddings

Hope this finally works. Let me know if you give it a try

File: llama.cpp (hunk 5 above)
Comment (ggerganov) [reply llama.cpp]: > Also fp32 is down from 9.34e-11 to 1.18e-14.

This is likely due to no longer going through the KV cache.

Do you have any performance benchmarks to see if this change improved the speed?

**Comment by iamlemec** [comment]
Working great here! Numbers look very, very similar to earlier.

I'm a little surprised to see a slight performance degredation (around 20-30% on both CPU and CUDA). We're definitely getting gains from the kv-cache attention matrix stuff when that's a factor. But I wonder if it relates to the fact that we're breaking up the final copy into sequences? Either that or one of the smaller changes in `built_bert`. Looking into it now.

File: examples/embedding/embedding.cpp
Comment (cebtenzzre) [inline examples/embedding/embedding.cpp:1]: Does the llama_kv_cache_clear still do anything useful?

**Comment by ggerganov** [comment]
> I'm a little surprised to see a slight performance degredation (around 20-30% on both CPU and CUDA). We're definitely getting gains from the kv-cache attention matrix stuff when that's a factor. But I wonder if it relates to the fact that we're breaking up the final copy into sequences? Either that or one of the smaller changes in built_bert. Looking into it now.

Let me know what you find. I guess the embeddings extraction is now indeed a bit slower due to the `std::map` and loop over the batch. The actual computation seems to be the same speed as before (at least with Metal that is). You can bench it with `llama-bench` since it does not involve any embeddings extraction.

The main struggle with the API was to find a way to differentiate between token embeddings and sequence embeddings. Without the `std::map` there was an ambiguity in which embeddings you were extracting. There might be better way to do it though

The big benefit of not using the KV cache is that now we don't need to allocate a large KV cache memory buffer. For example, you can compute 32k token embeddings with just a large enough batch size (e.g. `-c 512 -b 32768`)

File: examples/embedding/embedding.cpp
Comment (ggerganov) [reply examples/embedding/embedding.cpp:1]: Yes, `llama_kv_cache_clear` is not necessary for embedding models, but makes sense for causal attention models

**Comment by iamlemec** [comment]
Oh nice, I didn't know about `llama-bench`. So for CPU I'm seeing ~20% performance drop, not too bad. But for CUDA I'm getting an almost 60% drop:
```
| model          | backend    | ngl | test       |             (pre) t/s |            (post) t/s |
| -------------- | ---------- | --: | ---------- | --------------------: | ---------------------:|
| bert 109M F16  | CUDA       |  99 | pp 512     | 177281.98 ± 116298.41 |  74489.67 ±  19758.33 |
| bert 109M F16  | CUDA       |   0 | pp 512     |   6824.48 ±   1151.24 |   5556.47 ±    536.14 |
```
It seems like the kv cache code relies more on `ggml_view_3d` instead of contiguous permutes, but I haven't yet found anything that yields the old performance.
